### ISSUE TITLE
Debug::dump occasionally outputs additional value in array output

### ISSUE THREAD

**Issue by acoulton** [kohana/core#559]
The tests for Debug::dump occasionally fail with:

```
There were 3 failures:
1) Kohana_DebugTest::test_var with data set #0 (array('foobar'), '<pre class="debug"><small>array</small><span>(1)</span> <span>(
    0 => <small>string</small><span>(6)</span> "foobar"
)</span></pre>')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
 '<pre class="debug"><small>array</small><span>(1)</span> <span>(
     0 => <small>string</small><span>(6)</span> "foobar"
+    5425148641218 => <small>bool</small> TRUE
 )</span></pre>'
/home/travis/build/kohana/core/tests/kohana/DebugTest.php:46
/home/travis/build/kohana/core/vendor/kohana/unittest/classes/Kohana/Unittest/TestSuite.php:51
2) Kohana_DebugTest::test_dump with data set #4 (array('foobar'), 128, 10, '<small>array</small><span>(1)</span> <span>(
    0 => <small>string</small><span>(6)</span> "foobar"
)</span>')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
 '<small>array</small><span>(1)</span> <span>(
     0 => <small>string</small><span>(6)</span> "foobar"
+    5425148641218 => <small>bool</small> TRUE
 )</span>'
/home/travis/build/kohana/core/tests/kohana/DebugTest.php:124
/home/travis/build/kohana/core/vendor/kohana/unittest/classes/Kohana/Unittest/TestSuite.php:51
3) Kohana_DebugTest::test_dump with data set #7 (array(array(array(array(array('something'))))), 128, 4, '<small>array</small><span>(1)</span> <span>(
    "level1" => <small>array</small><span>(1)</span> <span>(
        "level2" => <small>array</small><span>(1)</span> <span>(
            "level3" => <small>array</small><span>(1)</span> <span>(
                "level4" => <small>array</small><span>(1)</span> (
                    ...
                )
            )</span>
        )</span>
    )</span>
)</span>')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
                 )
+                5425148641218 => <small>bool</small> TRUE
             )</span>
+            5425148641218 => <small>bool</small> TRUE
         )</span>
+        5425148641218 => <small>bool</small> TRUE
     )</span>
+    5425148641218 => <small>bool</small> TRUE
 )</span>'
/home/travis/build/kohana/core/tests/kohana/DebugTest.php:124
/home/travis/build/kohana/core/vendor/kohana/unittest/classes/Kohana/Unittest/TestSuite.php:51
```

These extra elements appear to be the `$marker` elements added in [Kohana_Debug:152](https://github.com/kohana/core/blob/3.3/develop/classes/Kohana/Debug.php#L152) to detect array recursion which for some reason aren't being skipped.

It happens quite rarely, and so far have only seen on PHP >= 5.4.

**Comment by enov** [comment]
It seems the problem relies on the way we are generating the unique key:
https://github.com/kohana/core/blob/3.3/develop/classes/Kohana/Debug.php#L137

I did this [small test](http://3v4l.org/0P9E8) and got different results on different PHP versions:

<pre>
$my_key = uniqid("\x00"); 
$var[$my_key] = TRUE;
foreach($var as $key => $val) 
{
    if ($key === $my_key) {
        echo 'You got it right';
    } else {
        echo 'You got it wrong';
    }
}
</pre>

I also [ran the same test again](http://3v4l.org/KFrtQ), with conforming results...

**Comment by acoulton** [comment]
@enov thanks, that set me thinking. Turns out the issue is that for certain clock values uniqid returns a string with all numeric characters. That becomes an integer key in the array, so the `===` will never pass for those times. There are some time periods when it never happens (eg now, your example passes at the mo) and then others where it happens depending on the microsecond.

I ran [a test that forces the key to be numeric](http://3v4l.org/OdZ2R) and [a test that forces it to be alpha](http://3v4l.org/CDnJN) and got consistent pass/fail on all platforms. Will commit a fix just now.
